File: app/src/main/java/me/devsaki/hentoid/fragments/library/LibraryBottomGroupsFragment.java
package me.devsaki.hentoid.fragments.library;

import static me.devsaki.hentoid.util.Preferences.Constant.ARTIST_GROUP_VISIBILITY_ARTISTS;
import static me.devsaki.hentoid.util.Preferences.Constant.ARTIST_GROUP_VISIBILITY_ARTISTS_GROUPS;
import static me.devsaki.hentoid.util.Preferences.Constant.ARTIST_GROUP_VISIBILITY_GROUPS;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.lifecycle.ViewModelProvider;

import com.google.android.material.bottomsheet.BottomSheetDialogFragment;
import com.mikepenz.fastadapter.FastAdapter;
import com.mikepenz.fastadapter.adapters.ItemAdapter;
import com.mikepenz.fastadapter.select.SelectExtension;

import java.util.ArrayList;
import java.util.List;

import me.devsaki.hentoid.R;
import me.devsaki.hentoid.activities.bundles.LibraryBottomSortFilterBundle;
import me.devsaki.hentoid.databinding.IncludeLibraryGroupsBottomPanelBinding;
import me.devsaki.hentoid.enums.Grouping;
import me.devsaki.hentoid.util.Preferences;
import me.devsaki.hentoid.util.ThemeHelper;
import me.devsaki.hentoid.viewholders.TextItem;
import me.devsaki.hentoid.viewmodels.LibraryViewModel;
import me.devsaki.hentoid.viewmodels.ViewModelFactory;

public class LibraryBottomGroupsFragment extends BottomSheetDialogFragment {

    private LibraryViewModel viewModel;

    // UI
    private IncludeLibraryGroupsBottomPanelBinding binding = null;

    // RecyclerView controls
    private final ItemAdapter<TextItem<Integer>> itemAdapter = new ItemAdapter<>();
    private final FastAdapter<TextItem<Integer>> fastAdapter = FastAdapter.with(itemAdapter);
    private SelectExtension<TextItem<Integer>> selectExtension;

    // Variables
    private boolean isCustomGroupingAvailable;

    public static synchronized void invoke(
            Context context,
            FragmentManager fragmentManager) {
        // Don't re-create it if already shown
        for (Fragment fragment : fragmentManager.getFragments())
            if (fragment instanceof LibraryBottomGroupsFragment) return;

        LibraryBottomSortFilterBundle builder = new LibraryBottomSortFilterBundle();
        LibraryBottomGroupsFragment libraryBottomSheetFragment = new LibraryBottomGroupsFragment();
        libraryBottomSheetFragment.setArguments(builder.getBundle());
        ThemeHelper.setStyle(context, libraryBottomSheetFragment, STYLE_NORMAL, R.style.Theme_Light_BottomSheetDialog);
        libraryBottomSheetFragment.show(fragmentManager, "libraryBottomSheetFragment");
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);

        Bundle bundle = getArguments();
        if (bundle != null) {
            LibraryBottomSortFilterBundle parser = new LibraryBottomSortFilterBundle(bundle);
        }

        ViewModelFactory vmFactory = new ViewModelFactory(requireActivity().getApplication());
        viewModel = new ViewModelProvider(requireActivity(), vmFactory).get(LibraryViewModel.class);

        viewModel.isCustomGroupingAvailable().observe(this, b -> {
            isCustomGroupingAvailable = b;
            itemAdapter.set(getGroupings());
        });
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = IncludeLibraryGroupsBottomPanelBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        // SORT TAB
        // Gets (or creates and attaches if not yet existing) the extension from the given `FastAdapter`
        selectExtension = fastAdapter.getOrCreateExtension(SelectExtension.class);
        if (selectExtension != null) {
            selectExtension.setSelectable(true);
            selectExtension.setMultiSelect(false);
            selectExtension.setSelectOnLongClick(false);
            selectExtension.setSelectWithItemUpdate(true);
            selectExtension.setAllowDeselection(false);
            selectExtension.setSelectionListener((item, selected) -> {
                if (selected) onSelectionChanged(item);
            });
        }
        binding.list.setAdapter(fastAdapter);

        updateArtistVisibility();

        binding.artistDisplayGrp.addOnButtonCheckedListener((g, i, b) -> {
            int code;
            if (binding.artistDisplayArtists.isChecked() && binding.artistDisplayGroups.isChecked())
                code = ARTIST_GROUP_VISIBILITY_ARTISTS_GROUPS;
            else if (binding.artistDisplayArtists.isChecked())
                code = ARTIST_GROUP_VISIBILITY_ARTISTS;
            else code = ARTIST_GROUP_VISIBILITY_GROUPS;
            Preferences.setArtistGroupVisibility(code);
            updateArtistVisibility();
            viewModel.searchGroup();
        });
    }

    private List<TextItem<Integer>> getGroupings() {
        List<TextItem<Integer>> result = new ArrayList<>();
        result.add(createFromGrouping(Grouping.FLAT));
        result.add(createFromGrouping(Grouping.ARTIST));
        result.add(createFromGrouping(Grouping.DL_DATE));
        if (isCustomGroupingAvailable) result.add(createFromGrouping(Grouping.CUSTOM));
        return result;
    }

    private TextItem<Integer> createFromGrouping(@NonNull Grouping grouping) {
        return new TextItem<>(
                getResources().getString(grouping.getName()),
                grouping.getId(),
                true,
                Preferences.getGroupingDisplay().getId() == grouping.getId());
    }

    private void updateArtistVisibility() {
        int visibility = (Preferences.getGroupingDisplay() == Grouping.ARTIST) ? View.VISIBLE : View.INVISIBLE;
        binding.artistDisplayTxt.setVisibility(visibility);
        binding.artistDisplayGrp.setVisibility(visibility);
        int code = Preferences.getArtistGroupVisibility();
        binding.artistDisplayArtists.setChecked(ARTIST_GROUP_VISIBILITY_ARTISTS_GROUPS == code || ARTIST_GROUP_VISIBILITY_ARTISTS == code);
        binding.artistDisplayGroups.setChecked(ARTIST_GROUP_VISIBILITY_ARTISTS_GROUPS == code || ARTIST_GROUP_VISIBILITY_GROUPS == code);
    }

    /**
     * Callback for any selected item
     */
    private void onSelectionChanged(TextItem<Integer> item) {
        Integer code = item.getTag();
        if (code != null) {
            viewModel.setGrouping(code);
            updateArtistVisibility();
        }
    }
}

File: app/src/main/java/me/devsaki/hentoid/fragments/library/LibraryBottomGroupsFragment.kt
package me.devsaki.hentoid.fragments.library

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.FragmentManager
import androidx.lifecycle.ViewModelProvider
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.mikepenz.fastadapter.FastAdapter
import com.mikepenz.fastadapter.ISelectionListener
import com.mikepenz.fastadapter.adapters.ItemAdapter
import com.mikepenz.fastadapter.select.SelectExtension
import me.devsaki.hentoid.R
import me.devsaki.hentoid.activities.bundles.LibraryBottomSortFilterBundle
import me.devsaki.hentoid.databinding.IncludeLibraryGroupsBottomPanelBinding
import me.devsaki.hentoid.enums.Grouping
import me.devsaki.hentoid.util.Settings
import me.devsaki.hentoid.util.setStyle
import me.devsaki.hentoid.viewholders.TextItem
import me.devsaki.hentoid.viewmodels.LibraryViewModel
import me.devsaki.hentoid.viewmodels.ViewModelFactory

class LibraryBottomGroupsFragment : BottomSheetDialogFragment() {
    private lateinit var viewModel: LibraryViewModel

    // UI
    private var binding: IncludeLibraryGroupsBottomPanelBinding? = null

    // RecyclerView controls
    private val itemAdapter = ItemAdapter<TextItem<Int>>()
    private val fastAdapter = FastAdapter.with(itemAdapter)
    private var selectExtension: SelectExtension<TextItem<Int>>? = null

    // Variables
    private var isCustomGroupingAvailable = false
    private var isDynamicGroupingAvailable = false

    companion object {
        @Synchronized
        fun invoke(
            context: Context,
            fragmentManager: FragmentManager
        ) {
            // Don't re-create it if already shown
            for (fragment in fragmentManager.fragments) if (fragment is LibraryBottomGroupsFragment) return
            val builder = LibraryBottomSortFilterBundle()
            val libraryBottomSheetFragment = LibraryBottomGroupsFragment()
            libraryBottomSheetFragment.arguments = builder.bundle
            context.setStyle(
                libraryBottomSheetFragment,
                STYLE_NORMAL,
                R.style.Theme_Light_BottomSheetDialog
            )
            libraryBottomSheetFragment.show(fragmentManager, "libraryBottomSheetFragment")
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        val vmFactory = ViewModelFactory(requireActivity().application)
        viewModel = ViewModelProvider(requireActivity(), vmFactory)[LibraryViewModel::class.java]
        viewModel.isCustomGroupingAvailable.observe(this) { b: Boolean ->
            isCustomGroupingAvailable = b
            itemAdapter.set(getGroupings())
        }
        viewModel.isDynamicGroupingAvailable.observe(this) { b: Boolean ->
            isDynamicGroupingAvailable = b
            itemAdapter.set(getGroupings())
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = IncludeLibraryGroupsBottomPanelBinding.inflate(inflater, container, false)
        return binding!!.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        // SORT TAB
        // Gets (or creates and attaches if not yet existing) the extension from the given `FastAdapter`
        selectExtension = fastAdapter.requireOrCreateExtension()
        selectExtension?.apply {
            isSelectable = true
            multiSelect = false
            selectOnLongClick = false
            selectWithItemUpdate = true
            allowDeselection = false
            selectionListener =
                object : ISelectionListener<TextItem<Int>> {
                    override fun onSelectionChanged(item: TextItem<Int>, selected: Boolean) {
                        if (selected) onSelectionChanged(item)
                    }
                }
        }
        binding?.apply {
            list.adapter = fastAdapter
            updateArtistVisibility()
            artistDisplayGrp.addOnButtonCheckedListener { _, _, _ ->
                val code =
                    if (artistDisplayArtists.isChecked && artistDisplayGroups.isChecked) Settings.Value.ARTIST_GROUP_VISIBILITY_ARTISTS_GROUPS
                    else if (artistDisplayArtists.isChecked) Settings.Value.ARTIST_GROUP_VISIBILITY_ARTISTS else Settings.Value.ARTIST_GROUP_VISIBILITY_GROUPS
                Settings.artistGroupVisibility = code
                updateArtistVisibility()
                viewModel.searchGroup()
            }
        }
    }

    private fun getGroupings(): List<TextItem<Int>> {
        val result: MutableList<TextItem<Int>> = ArrayList()
        result.add(createFromGrouping(Grouping.FLAT))
        result.add(createFromGrouping(Grouping.ARTIST))
        result.add(createFromGrouping(Grouping.DL_DATE))
        if (isDynamicGroupingAvailable) result.add(createFromGrouping(Grouping.DYNAMIC))
        if (isCustomGroupingAvailable) result.add(createFromGrouping(Grouping.CUSTOM))
        return result
    }

    private fun createFromGrouping(grouping: Grouping): TextItem<Int> {
        return TextItem(
            resources.getString(grouping.displayName),
            grouping.id,
            true,
            Settings.groupingDisplay == grouping.id
        )
    }

    private fun updateArtistVisibility() {
        binding?.apply {
            val visibility =
                if (Settings.getGroupingDisplayG() == Grouping.ARTIST) View.VISIBLE else View.INVISIBLE
            artistDisplayTxt.visibility = visibility
            artistDisplayGrp.visibility = visibility
            val code = Settings.artistGroupVisibility
            artistDisplayArtists.isChecked =
                Settings.Value.ARTIST_GROUP_VISIBILITY_ARTISTS_GROUPS == code || Settings.Value.ARTIST_GROUP_VISIBILITY_ARTISTS == code
            artistDisplayGroups.isChecked =
                Settings.Value.ARTIST_GROUP_VISIBILITY_ARTISTS_GROUPS == code || Settings.Value.ARTIST_GROUP_VISIBILITY_GROUPS == code
        }
    }

    /**
     * Callback for any selected item
     */
    private fun onSelectionChanged(item: TextItem<Int>) {
        item.getObject()?.let {
            viewModel.setGrouping(it)
            updateArtistVisibility()
        }
    }
}